doc/compiler/optimizer/IntroLoopOptimizations.md

```diff
@@ -0,0 +1,183 @@
+
+### Versioning with Multiple Loop Invariant Checks
+
+What if there are two bound checks: `a[n]` and `b[m]`? Both are invariant. Now we have four tests: `n < 0`, `n >= a.length`, `m < 0`, `m >= b.length`. If we create a version of the loop for every test, it will explode the code size. What happens in this case is that there is only one good version and there is only one bad version when Loop Versioner versions a particular loop. The good version is executed if all the conditions are satisfied for executing the good version of the loop, meaning none of the exception checks will fail. The bad loop runs if any one of the conditions fail. The bad version has all the checks left intact from the original loop. The assumption is that none of the checks are meant to fail. The likelihood of hitting the good loop is pretty good. This is how multiple checks are versioned with respect to the same loop.
+
+### Versioning with Well-behaved Loops with Variant Checks
+```
+for (i=0; i<n; i++) {
+   a[i] = i; // doing a bound check every time because “i" changes
+   }
+```
+This is a well-behaved loop: `n` and array `a` are invariant inside the loop; The index starts out with a value and increments by `1` every time. The range of `i` is known. As long as `i` is less than `n` to begin with, you will increment `i` until it becomes `n`, at which point, the loop exits.
+
+Since the loop is going to run `n` times, instead of checking `i < 0` and `i >= a.length` inside the loop in each iteration, you can check `i < 0` and `i+n >= a.length` in the loop preheader. The two edges/ends of the range for values of `i` are tested: the enter value `i` and the final value `i+n`, with respect to the two conditions. If the upper bound and the lower bound are guaranteed to be fine, the bound does not need to be checked inside the loop every time.
+
+If any of the conditions fail, you go to the slow loop. Somewhere down the line in the slow loop in some iteration, the bound check exception will be raised. That is good enough to send it to the slow loop and run all iterations in the slow loop.
+
+Therefore, with a well-behaved loop, you could use the starting value to do the less-than test and the ending/final value to do the greater-or-equal-to test. This allows you to eliminate the bound check even if the value changes within the loop.
+
+This trick cannot be used if `i` is not changed in a predictable way. For example, `i=foo()` and the bound check on `a[i]` inside the loop cannot be eliminated. The array itself cannot be changing within the loop either. If array `a` points to an array of `100` elements on the first iteration and points to an array of `1` element on the second iteration, the bound check cannot be eliminated inside the loop.  Another example is that if the index to the array is a field: such as `a[o.f]` and there is a call within the loop. That call could change the field `o.f`. In this case, the bound check on `a[o.f]` cannot be eliminated.
```

Comment:
> The array itself cannot be changing within the loop either.

This could be confusing to an untrained eye. "The array" could be interpreted as the values in the array, which is not what we're trying to say here. Instead I would say:

> The base array pointer/object cannot be changing within the loop either.